storage/innobase/btr/btr0sea.cc

```diff
@@ -167,21 +167,19 @@ will not guarantee success.
 @param[in]	index	index handler */
 static void btr_search_check_free_space_in_heap(const dict_index_t *index)
 {
-  /* Note that we peek the value of heap->free_block without reserving
-  the latch: this is ok, because we will not guarantee that there will
-  be enough free space in the hash table. */
-
-  buf_block_t *block= buf_block_alloc();
-  auto part= btr_search_sys.get_part(*index);
-
-  part->latch.wr_lock(SRW_LOCK_CALL);
```

Comment:
On a closer thought, we may need `rd_lock(SRW_LOCK_CALL)` here in order to prevent a race condition with `mem_heap_free(part->heap)` or `mem_heap_empty(part->heap)` (which I think should be covered by an exclusive latch). I am not sure if there can be such calls, so I must check this, and add a source code comment, with or without the latch acquisition, as appropriate.

Reply:
We indeed must check `btr_search_enabled` while holding `part->latch` in order to avoid reintroducing a race condition with a concurrent `btr_search_disable()`, which had been fixed in ad2bf1129cfa85c00072b46e0355fe14bf69ee54.